scipy/ndimage/morphology.py

```diff
@@ -1687,11 +1687,17 @@ def white_tophat(input, size=None, footprint=None, structure=None,
     if isinstance(output, numpy.ndarray):
         grey_dilation(tmp, size, footprint, structure, output, mode, cval,
                       origin)
-        return numpy.subtract(input, output, output)
+        if input.dtype == numpy.bool_:
+            return numpy.bitwise_xor(input, output, output)
+        else:
+            return numpy.subtract(input, output, output)
     else:
         tmp = grey_dilation(tmp, size, footprint, structure, None, mode,
                             cval, origin)
-        return input - tmp
+        if input.dtype == numpy.bool_:
```

Comment:
Having almost the same thing repeated twice is kind of ugly, and I think this can be refactored into something nicer, see if the following makes sense:


```
    tmp = grey_erosion(input, size, footprint, structure, None, mode,
                       cval, origin)
    tmp = grey_dilation(tmp, size, footprint, structure, output, mode, cval,
                      origin)
    if tmp is None:
        tmp = output

    if tmp.dtype == np.bool_ and input.dtype == np.bool_:
        np.bitwise_xor(input, tmp, out=tmp)
    else:
        np.subtract(input, tmp, out=tmp)

    return tmp
```

Why this should work (almost) the same:

- the check to see if output is a numpy array is not done in any other ndimage function: if it is either an array or `None`, things work fine, if it is something else an error will be raised. So we can call `grey_dilation` outside the branch passing `output` as the 5th argument without worrying much: you will get a previously not raised error if you make `output` something weird, like a list or a string, but that is likely a user mistake and it is good to warn them.
- ndimage has the crappy behaviour where, instead of always returning the `output` array, whether you pass it to the function or it gets allocated for you, it returns `None` if you give it an actual `output ` array to store results in, hence the need to check if `tmp` is None.
- At this point `tmp` is either the `output` array given by the user, or an array of the same dtype as `input` allocated by the call to `grey_dilation`. Either way, we can reuse it for our return value and pass it as a third argument to our ufunc of choice.